### PR TITLE
add health metrics

### PR DESCRIPTION
**Pull request by lxiaopei**

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -103,3 +112,43 @@ func main() {
...
 	}
 }
+
+// Function for fetching nsx health status and feeding it to the prometheus metric.
+func getHealthStatus(lastSeenStatus *string) error {
+    status := "healthy"
+	if err := nsxClient.NSXChecker.CheckNSXHealth(); err != nil {
+		status := "unhealthy"
+	}
+	// Return without updating the version if it stayed the same since last time.
```
Comment (dantingl) [inline cmd/main.go:122]: Why not return the same status again?

File: pkg/metrics/metrics.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	"net/http"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var (
+	log = logf.Log.WithName("metrics").WithName("metrics")
```
Comment (dantingl) [inline pkg/metrics/metrics.go:13]: One `WithName("metrics")` is enough?

File: pkg/metrics/metrics.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+// objects with Prometheus and installs the Prometheus http handler at the
+// default context.
+func RegisterMetricsAndInstallHandler(m *http.ServeMux, metricsAddr string) {
+	registry := prometheus.NewRegistry()
+	register(registry)
+	m.Handle("/metrics", promhttp.HandlerFor(registry, promhttp.HandlerOpts{ErrorHandling: promhttp.ContinueOnError}))
+	address := net.JoinHostPort("", metricsAddr)
+	http.ListenAndServe(address, m)
+	log.Info("Registered /metrics with prometheus")
+}
```
Comment (heypnus) [inline pkg/metrics/metrics.go:53]: kubebuilder has exposed the metrics port 8080 by default so I think we can reuse it. You can refer to https://book.kubebuilder.io/reference/metrics.html#publishing-additional-metrics

Comment (lxiaopei) [reply pkg/metrics/metrics.go]: thanks, Qian. So we just use the metrics server started by controller.

File: pkg/metrics/metrics.go (hunk 2 above)
Comment (lxiaopei) [reply pkg/metrics/metrics.go:12]: yes, removed the last one

File: cmd/main.go (hunk 1 above)
Comment (lxiaopei) [reply cmd/main.go:123]: if err happend, set the status to "unhealthy"

File: pkg/metrics/metrics.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+			Namespace: MetricNamespace,
+			Subsystem: MetricSubsystem,
+			Name:      HealthKey,
+			Help:      "Last health status for NSX-Operator. 1 for 'status' label with current status.",
+		},
+		[]string{"status"},
+	)
+)
+
+var registerMetrics sync.Once
```
Comment (heypnus) [inline pkg/metrics/metrics.go:34]: For `sync.Once`, I'm afraid we cannot invoke Register multiple times for different metrics. e.g.
```
metrics.Register(metrics.NSXOperatorHealthStats)
metrics.Register(metrics.NSXOperatorControllerSyncTotal)
metrics.Register(metrics.NSXOperatorControllerUpdateTotal)
...
```
will fail, right?

File: pkg/metrics/metrics.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import (
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+var (
+	log = logf.Log.WithName("metrics")
```
Comment (heypnus) [inline pkg/metrics/metrics.go:12]: It seems that `log` is unused in `metrics.go`, can we remove it?

Comment (lxiaopei) [reply pkg/metrics/metrics.go:12]: removed. thanks.

File: cmd/main.go [hunk 6]
```diff
@@ -33,8 +39,9 @@ func init() {
 }
 
 func main() {
-	var probeAddr string
+	var probeAddr, metricsAddr string
 	flag.StringVar(&probeAddr, "health-probe-bind-address", ":8384", "The address the probe endpoint binds to.")
+	flag.StringVar(&metricsAddr, "metrics-bind-address", ":8089", "The address the metrics endpoint binds to.")
```
Comment (heypnus) [inline cmd/main.go:44]: Please change the port to 8093 as we discussed.

Comment (lxiaopei) [reply cmd/main.go]: changed.

File: pkg/metrics/metrics.go (hunk 4 above)
Comment (lxiaopei) [reply pkg/metrics/metrics.go:35]: What error do you encounter with the above? I did not test it with multiple metrics though.
I checked the doc and think we can set it 
func InitializePrometheusMetrics() {
	log.Info("Initializing prometheus metrics")
	Register(NSXOperatorHealthStats, NSXOperatorControllerSyncTotal, NSXOperatorControllerUpdateTotal)
}

Comment (heypnus) [reply pkg/metrics/metrics.go:44]: got it, thanks.

**Review by heypnus — APPROVED** [review]

**Pull request merged**